File: backend/src/models/ChatMessage.js
const mongoose = require('mongoose');

const chatMessageSchema = new mongoose.Schema({
    joinRequestId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'JoinRequest',
        required: true
    },
    senderId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    message: {
        type: String,
        required: [true, 'Message cannot be empty'],
        trim: true,
        maxlength: [1000, 'Message cannot exceed 1000 characters']
    },
    isSystemMessage: {
        type: Boolean,
        default: false
    },
    isEdited: {
        type: Boolean,
        default: false
    }
}, {
    timestamps: true
});

// Index for fetching messages by join request
chatMessageSchema.index({ joinRequestId: 1, createdAt: 1 });

module.exports = mongoose.model('ChatMessage', chatMessageSchema);

File: app/lib/models/chat_message.dart
class ChatMessage {
  final String id;
  final String joinRequestId;
  final String senderId;
  final String message;
  final bool isSystemMessage;
  final DateTime createdAt;
  final bool isEdited;

  ChatMessage({
    required this.id,
    required this.joinRequestId,
    required this.senderId,
    required this.message,
    required this.isSystemMessage,
    required this.createdAt,
    this.isEdited = false,
  });

  factory ChatMessage.fromJson(Map<String, dynamic> json) {
    return ChatMessage(
      id: json['_id'] ?? json['id'],
      joinRequestId: json['joinRequestId'],
      senderId: json['senderId'],
      message: json['message'],
      isSystemMessage: json['isSystemMessage'] ?? false,
      createdAt: DateTime.parse(json['createdAt']),
      isEdited: json['isEdited'] ?? false,
    );
  }
}
